server/middleware/cached.go

```diff
@@ -0,0 +1,60 @@
+package middleware
+
+import (
+	"time"
+
+	"github.com/cloudwan/gohan/schema"
+	"github.com/patrickmn/go-cache"
+	"github.com/rackspace/gophercloud"
+)
+
+type CachedIdentityService struct {
+	inner IdentityService
+	cache *cache.Cache
+}
+
+func (c *CachedIdentityService) GetTenantID(tenantName string) (string, error) {
+	return c.inner.GetTenantID(tenantName)
+}
+
+func (c *CachedIdentityService) GetTenantName(tenantID string) (string, error) {
+	i, ok := c.cache.Get(tenantID)
+	if ok {
+		return i.(string), nil
+	}
+	name, err := c.inner.GetTenantName(tenantID)
+	if err != nil {
+		return "", err
+	}
+	err = c.cache.Add(tenantID, name, cache.DefaultExpiration)
```

Comment:
there's a race here. if the same token is received twice in a short time window, it'll be added twice to the cache. the first attempt will succeed, but the second will fail with an error. the error will be returned to the client as 401